<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Шаг-1.-Обзор-данных" data-toc-modified-id="Шаг-1.-Обзор-данных-0.1">Шаг 1. Обзор данных</a></span></li><li><span><a href="#Шаг-2.1-Заполнение-пропусков" data-toc-modified-id="Шаг-2.1-Заполнение-пропусков-0.2">Шаг 2.1 Заполнение пропусков</a></span></li><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления.-0.3">Шаг 2.2 Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных." data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных.-0.4">Шаг 2.3. Изменение типов данных.</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов." data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов.-0.5">Шаг 2.4. Удаление дубликатов.</a></span></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-0.6">Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода." data-toc-modified-id="Шаг-2.6.-Категоризация-дохода.-0.7">Шаг 2.6. Категоризация дохода.</a></span></li><li><span><a href="#Шаг-2.7.-Категоризация-целей-кредита." data-toc-modified-id="Шаг-2.7.-Категоризация-целей-кредита.-0.8">Шаг 2.7. Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-0.9">Ответы на вопросы.</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-0.9.0.1">Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-0.9.0.2">Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-0.9.0.3">Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-0.9.0.4">Вопрос 4: Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-1">Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков.

**Описание проекта и цель исследования**

Заказчик — кредитный отдел банка. Входные данные от банка — статистика о платёжеспособности клиентов.

Цель исследования — проверьте гипотезу: 

* Влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.



**Ход исследования**

Данные о поведении пользователей получаем из файла `data.csv`. Перед проверкой гипотезы делаем обзор данных с целью проверить качество данных. 

Проверяем данные на ошибки и аномалии и оцениваем их влияние на исследование. Затем, на этапе предобработки выбираем способ обработки аномальных значений и исправляем самые критичные ошибки данных. Обрабатываем пропуски и удаляем дубликаты. Корректируем тип данных в столбцах.

Проводим категоризацию данных по доходам и целям кредита для последующей проверки гипотезы.
 
Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Категоризация данных и проверка гипотезы.

### Шаг 1. Обзор данных

In [62]:
# импорт библиотеки pandas
import pandas as pd

In [63]:
# чтение файла с данными и сохранение в df
try:
    df = pd.read_csv('datasets/data.csv')
except:
    df = pd.read_csv('/datasets/data.csv')

In [64]:
# получение первых 10 строк таблицы df
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [65]:
# получение общей информации о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице двенадцать столбцов. Столбцы содержат различные типы данных.

Согласно документации к данным:
* `children` — количество детей в семье;
* `days_employed` — общий трудовой стаж в днях;  
* `dob_years` — возраст клиента в годах;
* `education` — уровень образования клиента;
* `education_id` — идентификатор уровня образования;
* `family_status` — семейное положение;
* `family_status_id` — идентификатор семейного положения;
* `gender` — пол клиента;
* `income_type` — тип занятости;
* `debt` — имел ли задолженность по возврату кредитов;
* `total_income` — ежемесячный доход;
* `purpose` — цель получения кредита;

В названиях колонок отсутствует нарушения стиля.

Количество значений в столбцах различается. Следовательно, в данных есть пропущенные значения, которые находятся в столбцах `days_employed` и `total_income`. Оба столбца содержат вещественные типы данных (float64).

**Выводы**

В каждой строке таблицы — данные о клиентах банка. Часть колонок характеризует физические признаки клиента — пол и возраст, другие относятся к его образованию, работе и семье. Остальные данные относятся к кредитной истории клиента: имеет ли он задолженность по возврату кредитов и цель получения кредита.

Предварительно можно утверждать, что, данных достаточно для проверки гипотезы. Однако встречаются пропуски в данных, которые нужно устранить и проверить данные на аномалии и наличие дубликатов.

### Шаг 2.1 Заполнение пропусков

In [66]:
# подсчёт пропусков
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [67]:
# вывод на экран строк с пропущенными значениями в столбце total_income
df[df['total_income'].isna()].head()

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

In [68]:
# подсчет доли пропущенных значений
missing_values_days = df['days_employed'].isna().sum()/len(df['days_employed'])
missing_values_income = df['total_income'].isna().sum()/len(df['total_income'])
print(f'Доля пропущеннных значений в столбце days_employed составляет: {missing_values_days:.1%}')
print(f'Доля пропущеннных значений в столбце total_income составляет: {missing_values_income:.1%}')

Доля пропущеннных значений в столбце days_employed составляет: 10.1%
Доля пропущеннных значений в столбце total_income составляет: 10.1%


In [1]:
# описательная статистика данных
print('Максимум:','{:.0f}'.format(df['total_income'].max()))
print('Минимум:','{:.0f}'.format(df['total_income'].min()))
print('Среднее арифметическое:','{:.0f}'.format(df['total_income'].mean()))
print('Медиана:','{:.0f}'.format(df['total_income'].median()))

NameError: name 'df' is not defined

In [70]:
# заполнение пропусков в столбце total_income медианным значением (в зависимости от типа занятости)
df['total_income'] = df['total_income'].fillna(df.groupby('income_type')['total_income'].transform("median"))
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

Все пропуски в столбце `total_income` заполнены медианным значением.

**Выводы**
* Кроме пропусков в столбце `days_employed` были обнаружены пропуски в столбце `total_income`. Все пропуски находятся в столбцах с количественными переменными и имеют тип `NaN`.
* Столбцы содержат одинаковое количество пропущенных значений. Доля пропусков в обоих столбцах составляет *10.1%*.
* Возможные причины появления пропусков в данных:   
   1) Пропуски в `total_income` не являются случайными — количество пропусков совпадает с количество пропусков в `days_employed`, следовательно пропуски в этих столбцах связаны. Возможной причиной может быть то, что потенциальный заёмщик не указал место работы и доход в анкете, либо сделал это неправильно.  
   2) Еще одной  причиной может быть сбой выгрузки данных из ПФР, либо внутренняя ошибка базы данных.  
   3) Также пропуски могут быть спецификой ведения базы данных — пока банк не получил документальное подтверждение дохода, выгрузка данных по трудовому стажу не происходит.
* В `total_income` содержится информация о доходах в виде количественных переменных (float64). Для данного типа данных лучший способ заполнения пропусков — медианные значения. Применение среднего значения привело бы к искажениям, а именно к завышению показателя, характеризующего выборку, т.к. данные по доходам имеют выбросы — максимальный доход составляет 2265604, минимальный — 20667,  когда как медиана — 145018.

### Шаг 2.2 Проверка данных на аномалии и исправления.

In [71]:
# вывод уникальных значений в столбцах
for column in df.columns:
    print()
    print(f'{column}:',df[column].unique())


children: [ 1  0  3  2 -1  4 20  5]

days_employed: [-8437.67302776 -4024.80375385 -5623.42261023 ... -2113.3468877
 -3112.4817052  -1984.50758853]

dob_years: [42 36 33 32 53 27 43 50 35 41 40 65 54 56 26 48 24 21 57 67 28 63 62 47
 34 68 25 31 30 20 49 37 45 61 64 44 52 46 23 38 39 51  0 59 29 60 55 58
 71 22 73 66 69 19 72 70 74 75]

education: ['высшее' 'среднее' 'Среднее' 'СРЕДНЕЕ' 'ВЫСШЕЕ' 'неоконченное высшее'
 'начальное' 'Высшее' 'НЕОКОНЧЕННОЕ ВЫСШЕЕ' 'Неоконченное высшее'
 'НАЧАЛЬНОЕ' 'Начальное' 'Ученая степень' 'УЧЕНАЯ СТЕПЕНЬ'
 'ученая степень']

education_id: [0 1 2 3 4]

family_status: ['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'Не женат / не замужем']

family_status_id: [0 1 2 3 4]

gender: ['F' 'M' 'XNA']

income_type: ['сотрудник' 'пенсионер' 'компаньон' 'госслужащий' 'безработный'
 'предприниматель' 'студент' 'в декрете']

debt: [0 1]

total_income: [253875.6394526  112080.01410244 145885.95229686 ...  89672.56115303
 244093.05050043  82047.

In [72]:
print('Максмальный трудовой стаж в годах', df['days_employed'].max()/365)
print('Минимальный трудовой стаж в годах', df['days_employed'].min()/365)
print()
print('Максмальный ежемесячный доход', df['total_income'].max())
print('Минимальный ежемесячный доход', df['total_income'].min())

Максмальный трудовой стаж в годах 1100.6997273296713
Минимальный трудовой стаж в годах -50.38068465909146

Максмальный ежемесячный доход 2265604.028722744
Минимальный ежемесячный доход 20667.26379327158


В следующих столбцах были обнаружены артефакты (аномалии):  
1) `children`:    
- отрицательное количество детей (-1);  
- неправдоподобно большое количество детей (20).  

2) `days_employed`:  
- отрицательное количество дней трудового стажа;
- огромные значения положительных значений трудового стажа — максимальное значение 1100 лет;
- вещественный тип данных;
- пропуски.
   
3) `dob_years`:    
- нулевые значения возраста клиента.  
   
4) `education`:    
- дубликаты в категориальных переменных.  

5) `family_status`:  
- различный регистр.  
   
6) `gender`:    
- нестандартные пропуски.  
   
7) `total_income`:    
- вещественный тип данных. 
      
8) `purpose`:    
- дубликаты. 

**Обработка артефактов столбца `children`**

In [73]:
# количество артефактов
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Всего в столбце 123 аномальных значения. Возможная причина артефактов — ошибка при вводе. Заменим значение -1 на 1, а значение 20 на 2.

In [74]:
df.loc[df['children'] == -1,'children'] = 1
df.loc[df['children'] == 20,'children'] = 2
print(df['children'].unique())

[1 0 3 2 4 5]


В столбце не осталось аномальных значений

**Обработка артефактов столбца `days_employed`**

Возможная причина огромных значений трудового стажа — нарушилась разрядность в значениях. Разделим положительные значения на 100 и получим значения одного порядка с отрицательными:

In [75]:
# нормирование положительных значений
df.loc[df['days_employed'] >0,'days_employed'] /= 100

In [76]:
# проверка порядка значений
print(df.loc[df['days_employed'] >0,'days_employed'].mean())
print(df.loc[df['days_employed'] <0,'days_employed'].mean())

3650.0430991626863
-2353.0159319988766


Теперь значения в столбце одного порядка.

Возможная причина отрицательных значений — перепутаны столбцы начала и окончания трудового стажа. Заменим отрицательные значения на положительные:

In [77]:
# подсчет количества отрицательных значений
df[df['days_employed'] <0]['days_employed'].count()

15906

In [78]:
# замена отрицательных значений на положительные
df.loc[df['days_employed'] <0,'days_employed'] = df.loc[df['days_employed'] <0,'days_employed'].abs()

In [79]:
# проверка замены значений
df[df['days_employed'] <0]['days_employed'].count()

0

Отрицательные значения заменены успешно.

In [80]:
# заполнение пропусков в столбце медианным значением
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Все пропуски в столбце заполнены медианным значением.

In [81]:
# замена вещественного типа данных в столбце на целочисленный
df['days_employed'] = df['days_employed'].astype('int64')

In [82]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0           3402         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [83]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  int64  
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


Теперь значения в `days_employed` корректы — отрицательные и аномально огромные значения отсутствуют, а пропуски заменены медианным значением. Вещественный тип данных в столбце заменен на целочисленный, а количество значений совпадает с первоначальным — все операции проведены успешно.

**Обработка артефактов столбца `dob_years`**

In [84]:
# количество артефактов
df[df['dob_years'] == 0]['dob_years'].count()

101

Возможная причина нулевых значений возраста клиента — ошибка при вводе или загрузке данных. Заменим нулевые значения на медианные для групп клиентов по типу занятости:

In [85]:
# расчет медианного возраста для клиентов сгруппированных по типу занятости
dob_years_med = df[df['dob_years'] != 0].groupby('income_type')['dob_years'].median().astype('int')
dob_years_med

income_type
безработный        38
в декрете          39
госслужащий        40
компаньон          39
пенсионер          60
предприниматель    42
сотрудник          39
студент            22
Name: dob_years, dtype: int64

In [86]:
# замена нулевых значений на медианные
for types,med in dob_years_med.items():
    df.loc[(df['dob_years'] == 0) & (df['income_type'] == types), 'dob_years'] = med

In [87]:
# проверка количества артефактов
df[df['dob_years'] == 0]['dob_years'].count()

0

Все нулевые значения в столбце `dob_years` заполнены соответствующим медианным значением.

**Обработка артефактов столбца `family_status`**

Возможная причина появления переменной с отличающимся регистром — ошибка ручного ввода. Приведем категориальные переменные к единому регистру:

In [88]:
# приведение к нижнему регистру
df['family_status'] = df['family_status'].str.lower()

In [89]:
# проверка регистра
df.sort_values(by='family_status')['family_status'].unique()

array(['в разводе', 'вдовец / вдова', 'гражданский брак',
       'женат / замужем', 'не женат / не замужем'], dtype=object)

**Обработка артефактов столбца `gender`**

In [90]:
# количество артефактов
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Строка с пропуском только одна — удалим её.

In [91]:
# удаление пропуска из датафрейма
df = df[df['gender'] != "XNA"]
print(df['gender'].value_counts())
df.shape

F    14236
M     7288
Name: gender, dtype: int64


(21524, 12)

Строка удалена. Оставшее количество строк в датафрейме: 21524

**Выводы**  

На этапе предобработки данных были проведены следующие операции:
- Были обработаны основные артефакты в столбцах `children`, `days_employed`;
- Все пропуски в столбце `days_employed` заполнены медианным значением;
- Нулевые данные в столбце `dob_years` заполненны медианным значением согласно группировке клиентов по типу занятости;
- Категориальные переменные в `family_status` приведены к нижнему регистру;
- Удалена единственная строка с пропуском в столбце `gender`

Так как столбец `purpose` заполняется в свободной форме, на данном этапе не будем приводить его к унифицированному виду, вместо этого далее классифицируем цели кредита в отдельном столбце.


### Шаг 2.3. Изменение типов данных.

In [92]:
# изменение типа данных на целочисленный
df['total_income'] = df['total_income'].astype('int64')

In [93]:
# проверка изменения типа данных
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21524 non-null  int64 
 1   days_employed     21524 non-null  int64 
 2   dob_years         21524 non-null  int64 
 3   education         21524 non-null  object
 4   education_id      21524 non-null  int64 
 5   family_status     21524 non-null  object
 6   family_status_id  21524 non-null  int64 
 7   gender            21524 non-null  object
 8   income_type       21524 non-null  object
 9   debt              21524 non-null  int64 
 10  total_income      21524 non-null  int64 
 11  purpose           21524 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.1+ MB


Тип данных в столбце `total_income` изменен на целочисленный

### Шаг 2.4. Удаление дубликатов.

Подсчитаем и обработаем явные дубликаты в данных.

In [94]:
# подсчет явных дубликатов
df.duplicated().sum()

54

In [95]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index(drop=True)

In [96]:
# проверка на отсутствие дубликатов
df.duplicated().sum()

0

На этапе проверки данных неявные дубликаты были обнаружены в столбце `education`. Обработаем их.

In [97]:
# просмотр уникальных названий в столбце `education`
df.sort_values(by='education')['education'].unique()

array(['ВЫСШЕЕ', 'Высшее', 'НАЧАЛЬНОЕ', 'НЕОКОНЧЕННОЕ ВЫСШЕЕ',
       'Начальное', 'Неоконченное высшее', 'СРЕДНЕЕ', 'Среднее',
       'УЧЕНАЯ СТЕПЕНЬ', 'Ученая степень', 'высшее', 'начальное',
       'неоконченное высшее', 'среднее', 'ученая степень'], dtype=object)

In [98]:
# приведение к нижнему регистру
df['education'] = df['education'].str.lower()

In [99]:
# проверка регистра
df.sort_values(by='education')['education'].unique()

array(['высшее', 'начальное', 'неоконченное высшее', 'среднее',
       'ученая степень'], dtype=object)

**Выводы** 
- Явные дубликаты в данных были удалены методом `drop_duplicates`, который удаляет индентичные строки. Неявные дубликаты в столбце `education` были удалены путем приведения данных к нижнему регистру.
- Возможными причинами появления дубликатов могут быть как технические ошибки так и человеческий фактор.

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создадим новый датафрейм, в котором каждому уникальному значению из `education` соответствует уникальное значение `education_id` 

In [100]:
# создание нового датафрейма
education_dict = df[['education','education_id']]

In [101]:
# удаление дубликатов из нового датафрейма
education_dict = education_dict.drop_duplicates().reset_index(drop=True)
education_dict

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

Создадим новый датафрейм, в котором каждому уникальному значению из `family_status` соответствует уникальное значение `family_status_id`

In [102]:
# создание нового датафрейма
family_status_dict = df[['family_status','family_status_id']]

In [103]:
# удаление дубликатов из нового датафрейма
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
family_status_dict

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

Удалим из исходного датафрейма столбцы `education` и `family_status`.

In [104]:
# удаление столбцов
df = df[['children', 'days_employed', 'dob_years', 'education_id',
       'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose']]

Удалим из исходного датафрейма явные дубликаты, которые образовались после удаления столбцов.

In [105]:
# подсчет явных дубликатов
df.duplicated().sum()

17

In [106]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых) и проверка на отсутствие дубликатов
df = df.drop_duplicates().reset_index(drop=True)
df.duplicated().sum()

0

In [107]:
# проверка нового датафрейма
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           3402         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

### Шаг 2.6. Категоризация дохода.

Создадим столбец `total_income_category` с категориями:

0–30000 — `E`;\
30001–50000 — `D`;\
50001–200000 — `C`;\
200001–1000000 — `B`;\
1000001 и выше — `A`.  


Для этого напишем функцию категоризации клиентов по уровню дохода

In [108]:
# функция категоризации клиентов по уровню дохода
def income_sorting(income):
    if income <= 30000:
        return 'E'
    if 30001 <= income <= 50000:
        return 'D'
    if 50001 <= income <= 200000:
        return 'C'
    if 200001 <= income <= 1000000:
        return 'B'
    else:
        return 'A'

In [109]:
# применение функции к столбцу 'total_income'
df['total_income_category'] = df['total_income'].apply(income_sorting)

In [110]:
# проверка работы функции
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           3402         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

### Шаг 2.7. Категоризация целей кредита.

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории: 

 `операции с автомобилем`;\
 `операции с недвижимостью`;\
 `проведение свадьбы`;\
 `получение образования`.  

In [111]:
# изучение данных в столбце 'purpose'
df.sort_values(by='purpose')['purpose'].unique()

array(['автомобили', 'автомобиль', 'высшее образование',
       'дополнительное образование', 'жилье',
       'заняться высшим образованием', 'заняться образованием',
       'на покупку автомобиля', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля', 'на проведение свадьбы',
       'недвижимость', 'образование', 'операции с жильем',
       'операции с коммерческой недвижимостью',
       'операции с недвижимостью', 'операции со своей недвижимостью',
       'покупка жилой недвижимости', 'покупка жилья',
       'покупка жилья для сдачи', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости', 'покупка недвижимости',
       'покупка своего жилья', 'получение высшего образования',
       'получение дополнительного образования', 'получение образования',
       'приобретение автомобиля', 'профильное образование',
       'ремонт жилью', 'свадьба', 'свой автомобиль',
       'сделка с автомобилем', 'сделка с подержанным автомобилем',
       'строительство 

In [112]:
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           3402         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

In [113]:
# функция категоризации цели получения кредита
def purpose_sorting(purpose):
    if  'авто' in purpose:
        return 'операции с автомобилем'
    elif 'недвиж' in purpose or 'жиль' in purpose:
        return 'операции с недвижимостью'
    elif 'свадьб' in purpose:
        return 'проведение свадьбы'
    elif 'образо' in purpose:
        return 'получение образования'
    else:
        return 'unknown'

In [114]:
# применение функции к столбцу 'purpose'
df['purpose_category'] = df['purpose'].apply(purpose_sorting)

In [115]:
# проверка работы функции
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           3402         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

### Ответы на вопросы.

##### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для ответа на вопрос создадим сводную таблицу количества клиентов-должников сгруппированную по количеству детей. Посчитаем % должников по каждой группе. Если с ростом количества детей доля должников уменьшается — мы можем выдвинуть гипотезу о наличии обратной взаимосвязи.

In [116]:
# создание сводной таблицы (sum - количество должников, count - количество клиентов) и расчет % должников по каждой группе
children_pivot = df.pivot_table(index=['children'], values='debt', aggfunc=['sum','count'])
children_pivot['ratio'] = children_pivot['sum']/children_pivot['count']*100
children_pivot

sum  count     ratio
          debt   debt          
children                       
0         1063  14090  7.544358
1          445   4855  9.165808
2          202   2128  9.492481
3           27    330  8.181818
4            4     41  9.756098
5            0      9  0.000000

**Вывод 1:** Анализ результатов показывает вариабельность доли должников среди групп клиентов с различным количеством детей — какой-либо тенденции не прослеживается. Можно утверждать, что скорее всего никакой связи между количеством детей и сроком возврата кредита нет.  
Однако, для дальнейшей проверки на значимость можно выдвинуть гипотезу о том, что отсутствие детей влияет на уменьшение срока возврата кредита, т.к. наблюдается некоторое снижение доли должников для группы клиентов не имеющей детей.

##### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

Для ответа на вопрос создадим сводную таблицу количества клиентов-должников сгруппированную в зависимости от семейного положения. Посчитаем % должников по каждой группе и оценим наличие взаимосвязи.

In [117]:
# получение таблицы с family_status
family_debt = df.merge(family_status_dict, on='family_status_id', how='left')
family_debt.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           3402         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category     family_status  
0                     B  операции с недвижимостью   женат / замужем  
1                     C    операции с автомобилем   женат / замужем  
2                     C  операции с недвижимостью   женат / замужем  
3                     B     получение образования   женат / замужем  
4                     C        проведение свадьбы  гражданский брак

In [118]:
# создание сводной таблицы (sum - количество должников, count - количество клиентов) и расчет % должников по каждой группе
family_pivot = family_debt.pivot_table(index=['family_status'], values='debt', aggfunc=['sum','count'])
family_pivot['ratio'] = family_pivot['sum']/family_pivot['count']*100
family_pivot

sum  count     ratio
                      debt   debt          
family_status                              
в разводе               85   1195  7.112971
вдовец / вдова          63    959  6.569343
гражданский брак       388   4150  9.349398
женат / замужем        931  12339  7.545182
не женат / не замужем  274   2810  9.750890

**Вывод 2:** Некоторую зависимость между семейным положением и возвратом кредита в срок можно проследить. Среди категорий `не женат/не замужем` и `гражданский брак` больше всего доля должников. В категории `вдовец/вдова` дела с возвратом обстоят лучше всего. Доля должников в категориях `в разводе` и `женат/замужем` примерно одинакова.

##### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Для ответа на вопрос создадим сводную таблицу количества клиентов-должников сгруппированную в зависимости от категории дохода. Посчитаем % должников по каждой группе и оценим наличие взаимосвязи.

In [119]:
# создание сводной таблицы (sum - количество должников, count - количество клиентов) и расчет % должников по каждой группе
income_pivot = df.pivot_table(index=['total_income_category'], values='debt', aggfunc=['sum','count'])
income_pivot['ratio'] = income_pivot['sum']/income_pivot['count']*100
income_pivot

sum  count     ratio
                       debt   debt          
total_income_category                       
A                         2     25  8.000000
B                       356   5041  7.062091
C                      1360  16015  8.492039
D                        21    350  6.000000
E                         2     22  9.090909

**Вывод 3:** Клиенты с наименьшим доходом (`E`) реже других выплачивают кредит в срок, однако группа `D` с доходом 30–50 тыс. руб./мес. имеет самый низкий % должников. По остальным группам показатель должников варьируется.

##### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

Для ответа на вопрос создадим сводную таблицу количества клиентов-должников сгруппированную в зависимости от цели кредита. Посчитаем % должников по каждой группе и оценим наличие взаимосвязи.

In [120]:
# создание сводной таблицы (sum - количество должников, count - количество клиентов) и расчет % должников по каждой группе
purpose_pivot = df.pivot_table(index=['purpose_category'], values='debt', aggfunc=['sum','count'])
purpose_pivot['ratio'] = purpose_pivot['sum']/purpose_pivot['count']*100
purpose_pivot

sum  count     ratio
                         debt   debt          
purpose_category                              
операции с автомобилем    403   4306  9.359034
операции с недвижимостью  782  10810  7.234043
получение образования     370   4013  9.220035
проведение свадьбы        186   2324  8.003442

**Вывод 4:**  
Клиенты охотнее выплачивают кредит если он взят под `операции с недвижимостью` или на `проведение свадьбы`. Процент должников по кредитам на `операции с автомобилем` и `получение образования` примерно одинаков.

## Общий вывод:

В ходе исследования были обработаны пропущенные значения, удалены дубликаты и проведена категоризация данных для целей кредита и дохода. С помощью сводных таблиц проведена оценка гипотезы о влиянии семейного положения и количества детей клиента на факт погашения кредита в срок.

Гипотеза частично подтвердилась:
* На погашение кредита в срок скорее влияет факт наличия или отсутствия детей, а не их количество;
* Среди категорий `не женат/не замужем` и `гражданский брак` больше всего доля должников. В категории `вдовец/вдова` доля погашения в срок наиболее высокая. Доля должников в категориях `в разводе` и `женат/замужем` примерно одинакова.

Выявлены следующие инсайты:
* Клиенты с наименьшим доходом (`E`) чаще других имеют задержки по погашению кредита в срок, однако группа `D` с доходом 30–50 тыс. руб./мес. имеет самый низкий процент должников. По остальным группам показатель должников варьируется.
* Клиенты охотнее выплачивают кредит если он взят под `операции с недвижимостью` или на `проведение свадьбы`. Процент должников по кредитам на `операции с автомобилем` и `получение образования` примерно одинаков.